crates/runtime/src/task_history.rs

```diff
@@ -0,0 +1,358 @@
+        self.cache_hit = Some(cache_hit);
+    }
+
+    pub fn set_error_message(&mut self, error_message: String) {
+        self.error_message = Some(error_message);
+    }
+
+    pub fn add_label(&mut self, key: String, value: String) {
+        self.labels.insert(key, value);
+    }
+
+    pub fn add_labels<I: IntoIterator<Item = (String, String)>>(&mut self, labels: I) {
+        self.labels.extend(labels);
+    }
+
+    pub async fn instantiate_table() -> Result<Arc<AcceleratedTable>, Error> {
+        let time_column = Some("start_time".to_string());
+        let time_format = Some(TimeFormat::UnixSeconds);
+
+        let retention = Retention::new(
+            time_column.clone(),
+            time_format,
+            Some(Duration::from_secs(24 * 60 * 60)), // 1 day
+            Some(Duration::from_secs(300)),
+            true,
+        );
+        let query_history_table_reference =
+            TableReference::partial(SPICE_RUNTIME_SCHEMA, DEFAULT_TASK_HISTORY_TABLE);
+        create_internal_accelerated_table(
+            query_history_table_reference,
+            Arc::new(TaskTracker::table_schema()),
+            Acceleration::default(),
+            Refresh::default(),
+            retention,
+            Arc::new(RwLock::new(Secrets::default())),
+        )
+        .await
+        .boxed()
+        .context(UnableToRegisterTableSnafu)
+    }
+
+    fn table_schema() -> Schema {
+        Schema::new(vec![
+            Field::new("id", DataType::Utf8, false),
+            Field::new("context_id", DataType::Utf8, false),
+            Field::new("parent_id", DataType::Utf8, true),
+            Field::new("task_type", DataType::Utf8, false),
+            Field::new("input_text", DataType::Utf8, false),
+            Field::new(
+                "start_time",
+                DataType::Timestamp(TimeUnit::Nanosecond, None),
+                false,
+            ), // Note: Used for time column of Retention
+            Field::new(
+                "end_time",
+                DataType::Timestamp(TimeUnit::Nanosecond, None),
+                false,
+            ),
+            Field::new("execution_time", DataType::Float32, false),
+            Field::new("outputs_produced", DataType::UInt64, false),
+            Field::new("cache_hit", DataType::Boolean, false),
+            Field::new("error_message", DataType::Utf8, true),
+            Field::new(
+                "labels",
+                DataType::Map(
+                    Arc::new(Field::new_struct(
+                        "entries",
+                        vec![
+                            Arc::new(Field::new("keys", DataType::Utf8, false)),
+                            Arc::new(Field::new("values", DataType::Utf8, false)),
+                        ],
+                        false,
+                    )),
+                    false,
+                ),
+                false,
+            ),
+        ])
+    }
+
+    pub async fn finish(mut self) {
```

Comment:
Maybe this shouldn't be async, and it does a `tokio::spawn` with the contents of this function so that it doesn't block callers.